src/pactServer.ts

```diff
@@ -0,0 +1,159 @@
+import childProcess, { ChildProcess, ExecOptions } from 'child_process'
+import fs from 'fs'
+import _ from 'lodash'
+import process from 'process'
+import tmp from 'tmp'
+import util from 'util'
+
+const writeFile = util.promisify(fs.writeFile)
+
+const SERVER_STARTUP_TIMEOUT_MS = 2000
+
+// Log lines that we expect to see when the Pact server starts up successfully.
+const SERVER_STARTUP_SUCCESS_LINES = [
+  'No replay found',
+  'starting on port',
+]
+
+// Log lines that indicate a failure on start up.
+const SERVER_STARTUP_FAILURE_LINES = [
+  'Address already in use',
+]
+
+export interface Killable {
+  kill: () => void
+}
+
+/**
+ * Helper function for executing a subprocess.
+ *
+ * Returns the ChildProcess object and a promise for the results of the subprocess.
+ */
+function spawn(args: string[], outFile: string, options: ExecOptions = {}): [Killable, Promise<string>] {
+  const cmd: string = args.concat([`>${outFile}`]).join(' ')
+
+  let child: ChildProcess
+  const resultPromise = new Promise<string>((resolve, reject) => {
+    child = childProcess.exec(cmd, options, (error, stdout, stderr) => {
+      if (error) {
+        reject(new Error(`${args[0]} exited with non-zero exit code ${error.code}.\nstderr: ${stderr}`))
+      }
+      resolve(stdout)
+    })
+  })
+
+  const processWrapper = {
+    kill: () => {
+      // The spawned process runs in a shell, so in addition to killing the shell, we have to
+      // add one to the pid to get the process itself.
+      // See e.g. https://github.com/nodejs/node/issues/2098
+      // I haven't found a cleaner way to handle this.
+      process.kill(child.pid)
+      process.kill(child.pid + 1)
+    }
+  }
+  return [processWrapper, resultPromise]
+}
+
+export default class PactServer {
+  private _killable?: Killable
+  private _url?: string
+  port: number
+  started: boolean
+
+  constructor(port: number) {
+    this.port = port
+    this.started = false
+  }
+
+  /**
+   * The localhost URL that the server is running on.
+   *
+   * Should succeed if called after awaiting start().
+   */
+  get url(): string {
+    if (this._url) {
+      return this._url
+    } else {
+      throw new Error('PactServer did not finish starting up.')
+    }
+  }
+
+  /**
+   * Run the Pact server in the background.
+   *
+   * Wait for the server to finish starting up before resolving and reject if start-up failed.
+   */
+  async start(onStartupLogs?: (logLines: string) => void): Promise<void> {
+    if (this.started) {
```

Comment:
`process.on('exit', function() {})` to kill child processes when the parent exits

Reply:
Not necessary in the new version

Reply:
Jk, it only exits if you exit with ^C. If you exit with `process.exit` or with an error the server will keep running in the background. Added your suggestion!